Warnings when compiling HPX 1.2.1 with gcc 9

## Expected Behavior

... Please describe the behavior you would have expected.

## Actual Behavior

```
BUILDSTDERR: /builddir/build/BUILD/hpx_1.2.1/hpx/lcos/detail/future_data.hpp:720:74:   required from 'hpx::lcos::detail::timed_future_data<Result>::timed_future_data(const time_point&, Result_&&) [with Result_ = const hpx::util::unused_type&; Result = void; std::chrono::_V2::steady_clock::time_point = std::chrono::time_point<std::chrono::_V2::steady_clock, std::chrono::duration<long int, std::ratio<1, 1000000000> > >]'
BUILDSTDERR: /builddir/build/BUILD/hpx_1.2.1/hpx/lcos/future.hpp:1740:65:   required from here
BUILDSTDERR: /builddir/build/BUILD/hpx_1.2.1/hpx/lcos/detail/future_data.hpp:721:17: warning: implicitly-declared 'constexpr hpx::util::unused_type::unused_type(const hpx::util::unused_type&)' is deprecated [-Wdeprecated-copy]
```
Full log: https://kojipkgs.fedoraproject.org//work/tasks/9834/32429834/build.log

## Steps to Reproduce the Problem


  1. Compile hpx 1.2.1 with gcc 9 and boost 1.69. Maybe, the boost version is not relevant for these warnings

## Specifications

... Please describe your environment

  - HPX Version: 1.2.1
  - Platform (compiler, OS): gcc 9, Fedora 30

--- a/hpx/error_code.hpp
+++ b/hpx/error_code.hpp
@@ -215,6 +215,12 @@ namespace hpx
             exception_ = std::exception_ptr();
         }
 
+        /// Copy constructor for error_code
+        ///
+        /// \note This function maintains the error category of the left hand
+        ///       side if the right hand side is a success code.
+        HPX_EXPORT error_code(error_code const& rhs);
+
         /// Assignment operator for error_code
         ///
         /// \note This function maintains the error category of the left hand

--- a/hpx/util/detail/pack_traversal_impl.hpp
+++ b/hpx/util/detail/pack_traversal_impl.hpp
@@ -280,7 +280,7 @@ namespace util {
             HPX_CONSTEXPR tuple<First, Rest...> voidify_empty_tuple(
                 tuple<First, Rest...> val)
             {
-                return std::move(val);
+                return val;
             }
             inline void voidify_empty_tuple(tuple<> const&) noexcept
             {

--- a/hpx/util/unused.hpp
+++ b/hpx/util/unused.hpp
@@ -1,6 +1,6 @@
 /*=============================================================================
     Copyright (c) 2001-2011 Joel de Guzman
-    Copyright (c) 2007-2013 Hartmut Kaiser
+    Copyright (c) 2007-2019 Hartmut Kaiser
 
     Distributed under the Boost Software License, Version 1.0. (See accompanying
     file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
@@ -23,37 +23,56 @@ namespace hpx { namespace util
     ///////////////////////////////////////////////////////////////////////////
     struct unused_type
     {
-        HPX_CONSTEXPR HPX_HOST_DEVICE unused_type() noexcept
+        HPX_CONSTEXPR HPX_HOST_DEVICE HPX_FORCEINLINE unused_type() noexcept {}
+
+        HPX_CONSTEXPR HPX_HOST_DEVICE HPX_FORCEINLINE unused_type(
+            unused_type const&)
+        {
+        }
+        HPX_CONSTEXPR HPX_HOST_DEVICE HPX_FORCEINLINE unused_type(unused_type&&)
         {
         }
 
         template <typename T>
-        HPX_CONSTEXPR HPX_HOST_DEVICE unused_type(T const&) noexcept
+        HPX_CONSTEXPR HPX_HOST_DEVICE HPX_FORCEINLINE unused_type(
+            T const&) noexcept
         {
         }
 
         template <typename T>
-        HPX_CONSTEXPR HPX_HOST_DEVICE unused_type const&
+        HPX_CONSTEXPR HPX_HOST_DEVICE HPX_FORCEINLINE unused_type const&
         operator=(T const&) const noexcept
         {
             return *this;
         }
 
         template <typename T>
-        HPX_HOST_DEVICE unused_type&
+        HPX_HOST_DEVICE HPX_FORCEINLINE unused_type&
         operator=(T const&) noexcept
         {
             return *this;
         }
 
-        HPX_CONSTEXPR HPX_HOST_DEVICE unused_type const&
+        HPX_CONSTEXPR HPX_HOST_DEVICE HPX_FORCEINLINE unused_type const&
         operator=(unused_type const&) const noexcept
         {
             return *this;
         }
 
-        HPX_HOST_DEVICE unused_type&
+        HPX_HOST_DEVICE HPX_FORCEINLINE unused_type&
         operator=(unused_type const&) noexcept
+        {
+            return *this;
+        }
+
+        HPX_CONSTEXPR HPX_HOST_DEVICE HPX_FORCEINLINE unused_type const&
+        operator=(unused_type&&) const noexcept
+        {
+            return *this;
+        }
+
+        HPX_HOST_DEVICE HPX_FORCEINLINE unused_type&
+        operator=(unused_type&&) noexcept
         {
             return *this;
         }

--- a/src/error_code.cpp
+++ b/src/error_code.cpp
@@ -176,6 +176,18 @@ namespace hpx
     }
 
     ///////////////////////////////////////////////////////////////////////////
+    error_code::error_code(error_code const& rhs)
+      : boost::system::error_code(rhs.value() == success ?
+                make_success_code(
+                    (category() == get_lightweight_hpx_category()) ?
+                        hpx::lightweight :
+                        hpx::plain) :
+                rhs)
+      , exception_(rhs.exception_)
+    {
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
     error_code& error_code::operator=(error_code const& rhs)
     {
         if (this != &rhs) {

